libs/parallelism/executors/include/hpx/executors/scheduler_executor.hpp

```diff
@@ -0,0 +1,319 @@
+                    std::forward<F>(f), std::move(t));
+            };
+        }
+
+        template <typename F, typename... Ts>
+        auto captured_args_unary(F&& f, Ts&&... ts)
+        {
+            return [f = std::forward<F>(f),
+                       t = hpx::make_tuple(std::forward<Ts>(ts)...)](
+                       auto&& arg) mutable {
+                return hpx::util::invoke_fused(
+                    hpx::util::bind_front(
+                        std::forward<F>(f), std::forward<decltype(arg)>(arg)),
+                    std::move(t));
+            };
+        }
+
+        template <typename F, typename... Ts>
+        auto captured_args_then(F&& f, Ts&&... ts)
+        {
+            return [f = std::forward<F>(f),
+                       t = hpx::make_tuple(std::forward<Ts>(ts)...)](
+                       auto i, auto&& predecessor, auto& v) mutable {
+                v[i] = hpx::util::invoke_fused(
+                    hpx::util::bind_front(std::forward<F>(f), i,
+                        std::forward<decltype(predecessor)>(predecessor)),
+                    std::move(t));
+            };
+        }
+
+        template <typename F, typename... Ts>
+        auto captured_args_then_void(F&& f, Ts&&... ts)
+        {
+            return [f = std::forward<F>(f),
+                       t = hpx::make_tuple(std::forward<Ts>(ts)...)](
+                       auto i, auto&& predecessor) mutable {
+                hpx::util::invoke_fused(
+                    hpx::util::bind_front(std::forward<F>(f), i,
+                        std::forward<decltype(predecessor)>(predecessor)),
+                    std::move(t));
+            };
+        }
+#endif
+    }    // namespace detail
+
+    ///////////////////////////////////////////////////////////////////////////
+    // A scheduler_executor wraps any P2300 scheduler and implements the
+    // executor functionalities for those.
+    template <typename BaseScheduler>
+    struct scheduler_executor
+    {
+        template <typename Scheduler,
+            typename Enable = std::enable_if_t<
+                hpx::execution::experimental::is_scheduler_v<Scheduler>>>
+        constexpr explicit scheduler_executor(Scheduler&& sched)
+          : sched_(std::forward<Scheduler>(sched))
+        {
+        }
+
+        /// \cond NOINTERNAL
+        constexpr bool operator==(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ == rhs.sched_;
+        }
+
+        constexpr bool operator!=(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ != rhs.sched_;
+        }
+
+        constexpr auto const& context() const noexcept
+        {
+            return *this;
+        }
+        /// \endcond
+
+        /// \cond NOINTERNAL
+        // Associate the parallel_execution_tag executor tag type as a default
+        // with this executor.
+        using execution_category = parallel_execution_tag;
```

Comment:
This is somewhat questionable, but probably the right default since the scheduler *can* run things in parallel but doesn't have to. We don't have a way to query the underlying scheduler if it will actually run `bulk` in parallel. Not really expecting a change here, but just wanted to point it out.

Reply:
Yes, I wasn't sure what to do here. However I believe we don't really use this tag for anything. We mostly rely on functions being available (or not). Will the schedulers expose their capabilities in any way?

Reply:
There's  currently nothing in P2300 similar to this tag. There have been questions about if `bulk` should support something like execution policies or if that should just be up to the scheduler (and not communicated outside of the scheduler), but no concensus as far as I can tell. I think it's rather far down on the list of open questions about senders/receivers at the moment.

I think the only place we rely on the tag is e.g. if you do `seq.on(exec)`, where the tag will be checked. `seq` requires `sequenced_execution_tag` whereas `par` requires either of the two. If `exec` is a `scheduler_executor` which only does a sequential `bulk_*` then `seq.on(exec)` will wrongly not compile, I think. However, I'm not too bothered by this.